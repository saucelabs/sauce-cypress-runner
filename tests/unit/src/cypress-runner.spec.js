jest.mock('cypress');
jest.mock('fs');
jest.mock('../../../src/sauce-reporter');
jest.mock('../../../src/utils');

const cypress = require('cypress');
const path = require('path');
const fs = require('fs');
const { sauceReporter, prepareAssets } = require('../../../src/sauce-reporter');
const { cypressRunner } = require('../../../src/cypress-runner');
const { loadRunConfig, getAbsolutePath } = require('../../../src/utils');

describe('.cypressRunner', function () {
  let oldEnv = { ...process.env };
  let cypressRunSpy;
  cypressRunSpy = jest.spyOn(cypress, 'run');
  beforeEach(function () {
    jest.resetModules();
    process.env = { ...oldEnv };
    cypressRunSpy.mockClear();
    const cypressRunResults = {
      runs: [{spec: {name: 'spec-a'}}, {spec: {name: 'spec-b'}}],
      failures: [],
    };
    cypress.run.mockImplementation(() => cypressRunResults);
    prepareAssets.mockClear();
    prepareAssets.mockImplementation(() => (['spec-a', 'spec-b']));
    const fakeRunnerJson = {
      cypress: {
        configFile: 'fake-cypress.json',
      },
      suites: [
        {name: 'fake-suite', config: {}}
      ]
    };
    getAbsolutePath.mockImplementation((path) => path);
    loadRunConfig.mockImplementation(() => fakeRunnerJson);
    fs.existsSync.mockImplementation(() => true);
    fs.readFileSync.mockImplementation(() => JSON.stringify(fakeRunnerJson));
  });
  it('can call Cypress.run with basic args', async function () {
    process.env.SAUCE_USERNAME = null;
    process.env.SAUCE_ACCESS_KEY = null;
    await cypressRunner('/fake/runner/path', 'fake-suite');
    // Change reporter to not be fully-qualified path
    const {reporter} = cypressRunSpy.mock.calls[0][0].config;
    cypressRunSpy.mock.calls[0][0].config.reporter = path.basename(reporter);
    expect(cypressRunSpy.mock.calls).toMatchSnapshot();
    expect(prepareAssets.mock.calls).toEqual([
      [
        ['spec-a', 'spec-b'], '/fake/runner/__assets__'
      ]
    ]);
  });
  it('can hardcode the browser path', async function () {
    process.env.SAUCE_BROWSER = 'C:/User/App/browser.exe:chrome';
    await cypressRunner('/fake/runner/path', 'fake-suite');
    const calledBrowser = cypressRunSpy.mock.calls[0][0].browser;
    expect(calledBrowser).toEqual('C:/User/App/browser.exe:chrome');
  });
  it('calls sauce reporter and returns a job status', async function () {
    process.env.SAUCE_USERNAME = 'bruno.alassia';
    process.env.SAUCE_ACCESS_KEY = 'i_l0ve_mayonnaise';
    process.env.SAUCE_BROWSER = 'firefox';
    sauceReporter.mockClear();
    await cypressRunner('/fake/runner/path', 'fake-suite');
    expect(sauceReporter.mock.calls).toMatchSnapshot();
  });
  it('throws error if browser is unsupported', function () {
    process.env.BROWSER_NAME = 'lynx';
    expect(cypressRunner('/fake/runner/path', 'fake-suite')).rejects.toThrow(new Error(
      `Unsupported browser: 'lynx'. Sorry.`
    ));
  });
  describe('from SAUCE VM', function () {
    it('returns false if there are test failures', async function () {
      process.env.SAUCE_VM = 'truthy';
      cypressRunSpy.mockImplementation(() => ({failures: 100}));
      const status = await cypressRunner('/fake/runner/path', 'fake-suite');
      expect(status).toEqual(false);
    });
    it('returns true if there are no test failures', async function () {
      process.env.SAUCE_VM = 'truthy';
      cypressRunSpy.mockImplementation(() => ({failures: 0}));
      const status = await cypressRunner('/fake/runner/path', 'fake-suite');
      expect(status).toEqual(false);
    });
  });
});