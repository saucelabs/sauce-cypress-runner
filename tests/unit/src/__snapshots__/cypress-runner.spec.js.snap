// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`.cypressRunner calls sauce reporter and returns a job status 1`] = `
Array [
  Array [
    Object {
      "cypress": Object {
        "configFile": "fake-cypress.json",
      },
      "path": "/fake/runner/path",
      "resultsDir": "/fake/runner/__assets__",
      "suites": Array [
        Object {
          "config": Object {},
          "name": "fake-suite",
        },
      ],
    },
    "fake-suite",
    "firefox",
    Array [
      "spec-a",
      "spec-b",
    ],
    Array [],
    "Date: 1",
    "Date: 2",
  ],
]
`;

exports[`.cypressRunner can call Cypress.run with basic args 1`] = `
Array [
  Array [
    Object {
      "browser": "chrome",
      "config": Object {
        "reporter": "custom-reporter.js",
        "reporterOptions": Object {
          "mochaFile": "/fake/runner/__assets__/[suite].xml",
          "specRoot": "cypress/integration",
        },
        "screenshotsFolder": "/fake/runner/__assets__",
        "testFiles": undefined,
        "video": undefined,
        "videoCompression": false,
        "videoUploadOnPasses": false,
        "videosFolder": "/fake/runner/__assets__",
      },
      "configFile": "fake-cypress.json",
      "project": "/fake/runner",
    },
  ],
]
`;
